feat: adding RLN support for nwaku-compose

gabrielmer opened the pull request:
# Description
<!--- Describe your changes to provide context for reviewrs -->
Added RLN support for nwaku-compose, so that RLN is ran by default.
Note that with this change, the user must manually add an Eth client WebSockets endpoint for nwaku-compose to work properly.

# Changes

<!-- List of detailed changes -->

- [x] Added RLN flags to `run_node.sh`
- [x] Defined volume for the RLN Tree in `docker-compose.yml` 
- [x] Added visualizations for RLN metrics in Grafana
- [x] Added support for optional RLN parameters `rln-relay-cred-password` and `rln-relay-cred-path`

Attached is an image of the resulting Grafana dashboard:

![image](https://github.com/waku-org/nwaku-compose/assets/101006718/e9bfce3d-7591-4360-a1f6-e6fb85634edb)

<!--
## How to test

1.
1.
1.

-->


## Issue

closes [#2051](https://github.com/waku-org/nwaku/issues/2051)

rymnc commented on docker-compose.yml:
```diff
@@ -10,8 +10,8 @@ x-pg-pass: &pg_pass ${POSTGRES_PASSWORD:-test123}
 x-pg-environment: &pg_env
-    POSTGRES_USER: *pg_user
-    POSTGRES_PASSWORD: *pg_pass
+  POSTGRES_USER: *pg_user
+  POSTGRES_PASSWORD: *pg_pass
```
is this change related to the pr?

vpavlin replied:
No, seems just like some auto-formatting fix by IDE

gabrielmer replied:
Exactly, it was the IDE's linter that removed the extra spaces.
Not related to the PR, if it creates noise I can remove it

rymnc reviewed (approved):
LGTM, one query

alrevuelta commented on run_node.sh:
```diff
@@ -66,6 +72,11 @@ exec /usr/bin/wakunode\
+  --rln-relay-dynamic=true\
+  --rln-relay-eth-contract-address="${RLN_RELAY_CONTRACT_ADDRESS}"\
+  --rln-relay-eth-client-address="${ETH_CLIENT_ADDRESS}"\
+  --rln-relay-tree-path="${RLN_TREE_PATH}"\
```
can we hardcode this? dont see much value on having it configurable. guess it should be `/etc/rln_tree` (the volume you are mounting)

gabrielmer replied:
Done :)

alrevuelta commented on run_node.sh:
```diff
@@ -66,6 +72,11 @@ exec /usr/bin/wakunode\
   --store-message-retention-policy=time:86400\
+  --rln-relay=true\
+  --rln-relay-dynamic=true\
+  --rln-relay-eth-contract-address="${RLN_RELAY_CONTRACT_ADDRESS}"\
```
can we hardcode it, or have a default value? one less parameter that a user has to configure

Ivansete-status replied:
I think we have a default value with the latest commits :100:

alrevuelta commented on run_node.sh:
```diff
@@ -41,6 +41,12 @@ if [ "${NODEKEY}" != "" ]; then
 fi
 
+RLN_RELAY_CONTRACT_ADDRESS="0xF471d71E9b1455bBF4b85d475afb9BB0954A29c4" # Sepolia Testnet
+RLN_TREE_PATH="/etc/rln_tree"
```
see other comment, lets remove this.

alrevuelta commented on run_node.sh:
```diff
@@ -41,6 +41,12 @@ if [ "${NODEKEY}" != "" ]; then
 
+RLN_RELAY_CONTRACT_ADDRESS="0xF471d71E9b1455bBF4b85d475afb9BB0954A29c4" # Sepolia Testnet
+RLN_TREE_PATH="/etc/rln_tree"
+ETH_CLIENT_ADDRESS= # Add a WebSockets endpoint to your Eth Node or an Infura WebSockets URL
```
not sure its needed to define it here? let remove this and explain the user that is has to EXPORT ETH_CLIENT_ADDRESS=BLABLA.

can we docuement this in the readme?

the idea is that this repo serves for non tech savy people, so ideally the should just do:

```
ETH_CLIENT_ADDRESS="ws://xxx"
docker-compose up -d
```


and everything should work

vpavlin commented:
A couple general notes:

* let's not defined **external** variables in `run_node.sh`, but rather in `docker-compose.yml`, so that we have one place where to look for the defaulf values
* Let's use defaults wherever possible - see https://github.com/waku-org/nwaku-compose/blob/master/docker-compose.yml#L9 or https://github.com/waku-org/nwaku-compose/blob/master/docker-compose.yml#L44 how to use env var with a default value
* I added a few notes in a comment on the issue before the PR was raised, so if we can consider those: https://github.com/waku-org/nwaku/issues/2051#issuecomment-1729166187
* As @alrevuelta mentioned, any optional or required nev vars should be primarily documented in README

In general, the goal is for `docker compose up` to work and get you to a usable node. Any additional configuration should be documented in README and ideally be as simple as possible (like for WSS, you only need to define `${DOMAIN}` and the rest is automated. Here it could be that you only need to provide blockchian WS RPC node URL and the rest is guessed/defaults are used, with ability to override values where necessary (but at that point you can also just "fork" the nwaku-compose repo and do what you need for your instance to work - but then we assume you are a power user:) )

gabrielmer commented:
Implemented new changes according to feedback.

Under the new commit:
- The path of the volume for the relay tree is hardcoded in `run_node.sh`
- `RLN_RELAY_CONTRACT_ADDRESS` and `ETH_CLIENT_ADDRESS` are now environment variables in `docker-compose.yml`. `RLN_RELAY_CONTRACT_ADDRESS` has default value `0xF471d71E9b1455bBF4b85d475afb9BB0954A29c4` and `ETH_CLIENT_ADDRESS` has no default value and must be provided by the user.

Updated README with information regarding the new environment variables.

Next points to improve:
1- Error handling for the case that the user does not provide an `ETH_CLIENT_ADDRESS`. I added an if condition on `run_node.sh` to check for its existence but the user can't see there is an error without looking specifically for the logs.
We discussed with @Ivansete-status that we should fail the whole `docker-compose` process in this case. Currently the nwaku container keeps trying to restart when `run_node.sh` fails.
If we go through this route, the modification would be done in a separate PR.

2- Better instructions in README to set up Infura node? I added a link to the most useful resource I found, in which it mentions how the Infura URL would look like depending on the protocol, network and user's API key.
Users might benefit more from a step by step guide of getting their Infura URL but not sure if the README is the right place for this guide. Please let me know what you think

3- As per @vpavlin comment, add support for the optional `rln-relay-cred-password` and
`rln-relay-cred-path` as env variables

I think that with the current changes + these 3 points mentioned, all the feedback proposals would be taken care of.
Please let me know if there's anything is missing, anything to be improved or if there's any misunderstanding from my end :)

Ivansete-status commented on README.md:
```diff
@@ -15,6 +15,20 @@ cd nwaku-compose
+It is required to set the `ETH_CLIENT_ADDRESS` environment variable before starting the instance.
+
+`ETH_CLIENT_ADDRESS` must be a WebSockets URL for your Ethereum Node.
+For the current default contract, it must a node for the Sepolia network.
```
```suggestion
For the current default contract, it must be a node for the Sepolia network.
```

gabrielmer replied:
Oh I missed it, will fix :)

Ivansete-status commented on README.md:
````diff
@@ -15,6 +15,20 @@ cd nwaku-compose
+For the current default contract, it must a node for the Sepolia network.
+In case you're not running your own node, you can get it from [Infura](https://www.infura.io/)
+
+You can set the environment variable from inside the `docker-compose.yml` file or you can simply run
+
+```console
+export ETH_CLIENT_ADDRESS=<your_eth_client_address_here>
+```
````
Better only mention modifying the env var in the docker-compose file, and then, create it as an anchor. Something like:

```
x-eth-client-address: &eth_client_address ${ETH_CLIENT_ADDRESS:-<YOUR_ETH_CLIENT_ADDRESS_HERE>}
```
As an example, check how `*pg_env` is used.

```suggestion
You can set the environment variable from inside the `docker-compose.yml` file.
```

gabrielmer replied:
Yes, I think that as an anchor is better! Will set it up as an anchor along with the optional RLN env variables that I'm about to add (`rln-relay-cred-password` and `rln-relay-cred-path`).

I mentioned the option of exporting the variable from the shell so it's more user friendly to some, as they don't need to look for and change any files in order to run their instance.
It was based on @alrevuelta idea of how the user experience should be, but not sure if it's redundant or if it might confuse users to mention two ways of doing the same thing.

Ivansete-status reviewed (commented):
Added a few comments :100:

gabrielmer commented on run_node.sh:
```diff
@@ -37,10 +42,18 @@ if [ -n "${DOMAIN}" ]; then
 fi
 
-if [ "${NODEKEY}" != "" ]; then
+if [ -n "${NODEKEY}" ]; then
```
Changed it for consistency with the other if conditions

Ivansete-status commented on README.md:
```diff
@@ -15,6 +15,12 @@ cd nwaku-compose
 
 Edit the environment variables present at the beginning of the `docker-compose.yml` file.
 
+It is required to set the `ETH_CLIENT_ADDRESS` environment variable before starting the instance.
```
Make sure it is edited within the "docker-compose.yml" file:

```suggestion
It is required to set the `ETH_CLIENT_ADDRESS` environment variable within the `docker-compose.yml` file before starting the instance.
```

gabrielmer replied:
I didn't specify it because one line above we instruct to edit the env variables in `docker-compose.yml`, so didn't want to repeat the same point made in the previous line.

What do you think? Is it better to repeat it?

Ivansete-status commented on run_node.sh:
```diff
@@ -66,7 +79,14 @@ exec /usr/bin/wakunode\
+  --rln-relay-eth-contract-address="${RLN_RELAY_CONTRACT_ADDRESS}"\
+  --rln-relay-eth-client-address="${ETH_CLIENT_ADDRESS}"\
+  --rln-relay-tree-path="/etc/rln_tree"\
   ${DNS_WSS_CMD}\
   ${NODEKEY}\
+  ${RLN_RELAY_CRED_PATH}\
+  ${RLN_RELAY_CRED_PASSWORD}\
```
Little nitpick. Should we have all the "rln" items together?

```suggestion
  ${RLN_RELAY_CRED_PATH}\
  ${RLN_RELAY_CRED_PASSWORD}\
  ${DNS_WSS_CMD}\
  ${NODEKEY}\
```

gabrielmer replied:
> Little nitpick. Should we have all the "rln" items together?

Oh yes! I saw the optional parameters as if there were in a different category, but with what we currently have we can change the order and have the RLN mandatory parameters be next to the RLN optional ones. Will fix it now

Ivansete-status commented on README.md:
```diff
@@ -36,4 +42,8 @@ There are multiple environment variables you can configure to modify behaviour o
 * `NWAKU_IMAGE` - the image you want to use for the nwaku container (e.g. `NWAKU_IMAGE=statusteam/nim-waku:v0.19.0-rc.0`)
 * `DOMAIN` - domain name pointing to the IP address of your node, when configured the run script will request SSL certs from Let's Encrypt and run Waku node with WebSockets Secure (WSS) options enabled (e.g. `DOMAIN=waku.example.com`)
 * `NODEKEY` - this env variable allows you to provide a node key as described in [operators documentation](https://github.com/waku-org/nwaku/blob/master/docs/operators/how-to/configure-key.md) (e.g. `NODEKEY=9f439983aa4851346cfe6e17585e426f482871a43626812e23490895cd602c11`)
+* `RLN_RELAY_CONTRACT_ADDRESS` - address of the RLN Relay Contract. It defaults to `0xF471d71E9b1455bBF4b85d475afb9BB0954A29c4` on Sepolia testnet
```
Small nitpick, avoid hard-coded values that we may need to update in the future. Is enough to only hard-code in the docker-compose file.

```suggestion
* `RLN_RELAY_CONTRACT_ADDRESS` - address of the RLN Relay Contract. It defaults to a Sepolia testnet address.
```

gabrielmer replied:
Done :)

Ivansete-status reviewed (commented):
Looks great so far!

alrevuelta reviewed (approved):
If its been tested and works and noone has more concerns, lgtm!

vpavlin reviewed (commented):
I am very concerned about merging this PR until we have a clear UX/DevEx about registering the membership.

IIUC enabling RLN in your node will basically prevent you from publishing a message unless you are a member of RLN contract - for which we have no straighforward way to happen right now.

I failed to convince @alrevuelta to make RLN optional for nwaku-compose for now, so the other option is to not merge the PR until we can clearly document how to obtain a membership.

My reasoning is that we have event(s) coming up where we want to showcase nwaku-compose and if people fail to use Waku due to RLN  they might be reluctant to try again in the future.

alrevuelta commented:
Lets block this PR from merge, until i) RLN membership registration process is done (via UI) and ii) the waku network is live with some rln nodes.
Added the blocked label.

gabrielmer commented:
As per the [documentation](https://rfc.vac.dev/spec/64/#memberships), modified the RLN contract address to `0x58322513A35a8f747AF5A385bA14C2AbE602AA59`

rymnc commented:
@gabrielmer the correct address is 0xF1935b338321013f11068abCafC548A7B0db732C, you've used the storage address

alrevuelta commented:
> 0xF1935b338321013f11068abCafC548A7B0db732C

indeed, my bad.

gabrielmer commented:
> @gabrielmer the correct address is 0xF1935b338321013f11068abCafC548A7B0db732C, you've used the storage address

Thanks for the clarification! Reverted the commit that changed it :)

alrevuelta commented:
@gabrielmer thanks! but not sure about the revert. Can we set?
0xF1935b338321013f11068abCafC548A7B0db732C

alrevuelta commented:
Merging since we already have an easy way to generate memberships, see https://github.com/waku-org/waku-frontend
